Makefile

```diff
@@ -1059,6 +1046,45 @@ ocamlyacc:
 clean::
 	$(MAKE) -C yacc clean
 
+# The Menhir-generated parser
+
+# In order to avoid a build-time dependency on Menhir,
+# we store the result of the parser generator (which
+# are OCaml source files) and Menhir's runtime libraries
+# (that the parser files rely on) in boot/
+
+parsing/parser.ml: \
+  boot/menhir/parser.ml parsing/parser.mly
+	@if [ parsing/parser.mly -nt boot/menhir/parser.ml ]; \
+	then \
+	  echo; \
+	  tput setaf 3; tput bold; printf "Warning: "; tput sgr0; \
+	  echo "Your 'parser.mly' file is more recent \
+	than the parser in 'boot/'."; \
+	  echo "Its changes will be ignored unless you run:"; \
+	  echo "    make promote-menhir"; \
+	  echo; \
```

Comment:
I just saw this warning fly by on my Git clone. I think the logic needs to be improved - there's no way of guaranteeing when you git clone (or untar a release tarball) that `boot/parser/parser.ml` will be newer than `parsing/parser.mly`

Would it be better to test for the presence of Menhir first?

Reply:
Testing for the presence of Menhir is not very informative: the user may modify parsing/parser.mly even if they don't have Menhir installed (for example, they are familiar with the older workflow), and they need the warning in either cases.

Maybe one way to strengthen the warning would be to test for a time delay between the two files that is longer than typical clone/untar operations incur. For example, It is going to be most often true that there is a gap of one minute between last touching `boot/` (for example on branch checkout) and manually modifying `parsing/parser.mly`.

Reply:
The `-nt` test operator is not described in POSIX shell&utilities and doesn't seem to work under MacOS.

Reply:
Back in the days when CompCert had a Menhir-generated parser checked in into the git repo, we did this:
* have a Makefile rule Parser.v: Parser.vy that calls Menhir
* at the end of configure, "touch Parser.v" so that it is more recent than Parser.vy and Menhir isn't called.

That was both fragile and source of unnecessary recompilations (of the generated Parser.v).  But that's the best we came up with.

Reply:
Checksums can also be used as a better alternative to timestamps:
* When running `menhir foo.mly`, also do `sha1sum foo.mly > foo.mly.chk`
* Warn when `sha1sum -c --status foo.mly.chk` fails.

Of course MacOS needs special handling (no `sha1sum` command but `shasum` might work), and I have no idea what the BSDs need.

Reply:
@xavierleroy I just pushed in a0d2aeb7e122d07ffe94f1956b4e056db571383a a solution using `find .. -prune -newer` instead of `-nt`, which should be POSIX; it might even work on macos.

I was enthusiastic about touching on configure as a way to make direct timestamp-comparison more robust, but it doesn't fix potential `git checkout` issues (because people typically don't reconfigure after they move to a different branch).

(In the current setting there is no `Parser.v: ...` rule that calls Menhir, because I wanted to make sure that Menhir would never be invoked by surprise during a build. You have to run `promote-menhir` explicitly.)